File: solutions/java/baffling-birthdays/1/src/main/java/BafflingBirthdays.java
import java.time.LocalDate;
import java.time.MonthDay;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Random;
import java.util.Set;

class BafflingBirthdays {
    private static final int TRIALS = 10000;
    private final Random random = new Random();

    boolean sharedBirthday(List<LocalDate> birthdates) {
        Set<MonthDay> seen = new HashSet<>();
        for (LocalDate date : birthdates) {
            // MonthDay extracts just the month and day, ignoring the year
            if (!seen.add(MonthDay.from(date))) {
                return true; // Found a duplicate birthday
            }
        }
        return false;
    }

    List<LocalDate> randomBirthdates(int groupSize) {
        List<LocalDate> dates = new ArrayList<>(groupSize);
        // Use a known non-leap year (2023) to strictly satisfy the 365-day assumption
        LocalDate startOfYear = LocalDate.of(2023, 1, 1);
        
        for (int i = 0; i < groupSize; i++) {
            // Uniformly pick a day from 1 to 365
            int dayOfYear = random.nextInt(365) + 1;
            dates.add(startOfYear.plusDays(dayOfYear - 1));
        }
        
        return dates;
    }

    double estimatedProbabilityOfSharedBirthday(int groupSize) {
        // Edge cases
        if (groupSize <= 1) {
            return 0.0;
        }
        // Pigeonhole principle: more than 365 people guarantees a shared birthday
        if (groupSize > 365) {
            return 100.0;
        }
        
        int sharedCount = 0;
        
        // Monte Carlo simulation
        for (int i = 0; i < TRIALS; i++) {
            if (sharedBirthday(randomBirthdates(groupSize))) {
                sharedCount++;
            }
        }
        
        // Return as a percentage (e.g., 50.7 for ~50.7%)
        return (double) sharedCount / TRIALS * 100.0;
    }
}

File: solutions/php/baffling-birthdays/1/BafflingBirthdays.php
<?php

/*
 * By adding type hints and enabling strict type checking, code can become
 * easier to read, self-documenting and reduce the number of potential bugs.
 * By default, type declarations are non-strict, which means they will attempt
 * to change the original type to match the type specified by the
 * type-declaration.
 *
 * In other words, if you pass a string to a function requiring a float,
 * it will attempt to convert the string value to a float.
 *
 * To enable strict mode, a single declare directive must be placed at the top
 * of the file.
 * This means that the strictness of typing is configured on a per-file basis.
 * This directive not only affects the type declarations of parameters, but also
 * a function's return type.
 *
 * For more info review the Concept on strict type checking in the PHP track
 * <link>.
 *
 * To disable strict typing, comment out the directive below.
 */

declare(strict_types=1);

class BafflingBirthdays
{
    /**
     * Checks if any two dates in the array share the same month and day.
     */
    public function sharedBirthday(array $birthdates): bool
    {
        if (count($birthdates) < 2) {
            return false;
        }

        $monthDays = [];
        foreach ($birthdates as $date) {
            // Handle both string dates and DateTime objects
            $dt = $date instanceof \DateTimeInterface ? $date : new \DateTime($date);
            $monthDays[] = $dt->format('m-d');
        }

        // If unique month-day count is less than total count, there's a duplicate
        return count($monthDays) !== count(array_unique($monthDays));
    }

    /**
     * Generates an array of random birthdates following the 365-day uniform distribution.
     */
    public function randomBirthdates(int $count): array
    {
        // Cache the 365 valid dates for a non-leap year to ensure uniform distribution
        static $validDates = null;
        if ($validDates === null) {
            $validDates = [];
            for ($month = 1; $month <= 12; $month++) {
                $daysInMonth = (int)date('t', mktime(0, 0, 0, $month, 1, 2023));
                for ($day = 1; $day <= $daysInMonth; $day++) {
                    $validDates[] = sprintf('2023-%02d-%02d', $month, $day);
                }
            }
        }

        $birthdates = [];
        for ($i = 0; $i < $count; $i++) {
            $birthdates[] = $validDates[array_rand($validDates)];
        }
        
        return $birthdates;
    }

    /**
     * Estimates the probability of a shared birthday using Monte Carlo simulation.
     * Returns probability as a percentage (0-100).
     */
    public function estimatedProbabilityOfSharedBirthday(int $groupSize): float
    {
        $simulations = 10000;
        $sharedCount = 0;

        for ($i = 0; $i < $simulations; $i++) {
            if ($this->sharedBirthday($this->randomBirthdates($groupSize))) {
                $sharedCount++;
            }
        }

        // Return as percentage (0-100), not decimal (0-1)
        return ($sharedCount / $simulations) * 100;
    }
}